src/Sylius/Bundle/CoreBundle/EventListener/ImagesRemoveListener.php

```diff
@@ -30,20 +34,34 @@
     /** @var FilterManager */
     private $filterManager;
 
+    /** @var string[] */
+    private $imagesToDelete = [];
+
     public function __construct(ImageUploaderInterface $imageUploader, CacheManager $cacheManager, FilterManager $filterManager)
     {
         $this->imageUploader = $imageUploader;
         $this->cacheManager = $cacheManager;
         $this->filterManager = $filterManager;
     }
 
-    public function postRemove(LifecycleEventArgs $event): void
+    public function onFlush(OnFlushEventArgs $event): void
     {
-        $image = $event->getEntity();
+        foreach ($event->getEntityManager()->getUnitOfWork()->getScheduledEntityDeletions() as $entityDeletion) {
+            if (!$entityDeletion instanceof ImageInterface) {
+                continue;
+            }
+
+            if (!in_array($entityDeletion->getPath(), $this->imagesToDelete)) {
+                $this->imagesToDelete[] = $entityDeletion->getPath();
+            }
+        }
+    }
 
-        if ($image instanceof ImageInterface) {
-            $this->imageUploader->remove($image->getPath());
-            $this->cacheManager->remove($image->getPath(), array_keys($this->filterManager->getFilterConfiguration()->all()));
+    public function postFlush(PostFlushEventArgs $event): void
+    {
+        foreach ($this->imagesToDelete as $imagePath) {
+            $this->imageUploader->remove($imagePath);
```

Comment:
We should also remove deleted images from `$this->imagesToDelete` array.

Reply:
An additional test for that should be nice also (PHP-PM community would be thankful) :)

Reply:
👍